### PR TITLE
Performance improvement

### PR DESCRIPTION
**Pull request by doskabouter**
First check envelope before calling CGAlgorithms::isPointInRing

### CODE REVIEW

**Comment by dbaston** [comment]
Thanks for this, @doskabouter . I haven't merged it yet because I want to verify / quantify the performance impact myself. (I did test your earlier PR through PostGIS and saw quite a large performance gain, but Postgres adds a lot of noise to performance testing. I'd like to try it by calling GEOS directly from a simple C program.)

**Comment by doskabouter** [comment]
And another 25% faster. Total improvement: twice as fast

File: include/geos/operation/overlay/PolygonBuilder.h [hunk 1]
```diff
@@ -152,6 +152,8 @@ class GEOS_DLL PolygonBuilder {
 		std::vector<geomgraph::EdgeRing*> &newShellList,
 		std::vector<geomgraph::EdgeRing*> &freeHoleList);
 
+    using FastPIPRing = tuple<EdgeRing*, algorithm::locate::IndexedPointInAreaLocator*>;
```
[the remark this answers is not in the document]
Comment (doskabouter) [reply include/geos/operation/overlay/PolygonBuilder.h]: internal struct sounds fine, I'll do that soon.
not using std::unique_ptr is due to my inexperience with c++. I assume that when used, the loop where those are delete'd can be removed?

[the remark this answers is not in the document]
Comment (dbaston) [reply include/geos/operation/overlay/PolygonBuilder.h]: >  I assume that when used, the loop where those are delete'd can be removed?

Yes, exactly

[the remark this answers is not in the document]
Comment (doskabouter) [reply include/geos/operation/overlay/PolygonBuilder.h]: Well, it's quite a hassle (in fact I didn't succeed at it yet), to have a std::vector of structs containing a std::unique_ptr, I keep getting compiler errors about "attempting to reference a deleted function".
I could probably fix this by adding appropriate copy-constructor/move-constructor/destrutors but that seems (at least to me at the moment) like quite the overkill.

[the remark this answers is not in the document]
Comment (dbaston) [reply include/geos/operation/overlay/PolygonBuilder.h]: Are you working with Visual Studio by chance? If so, you may need to explicitly delete the copy constructor and assignment operator; see https://github.com/libgeos/geos/blob/master/include/geos/index/strtree/SIRtree.h#L83

If you can't make it work, just leave the raw pointer.

[the remark this answers is not in the document]
Comment (doskabouter) [reply include/geos/operation/overlay/PolygonBuilder.h]: Indeed I'm using VS (2017). Must say those copy constructor issues are the least of the problems...
Main one being that VS doesn't resolve the namespaces correctly (I had to prepend a huge number of types with their namespace to get it to compile, and as such it's quite a hassle to create a proper pullrequesst,...)

**Comment by doskabouter** [comment]
Shall I continue adding improvements to this branch, (perhaps after this PR is merged?) or shall I create a new one for some other (f.e. missing envelope-checks) improvements?

**Comment by dbaston** [comment]
I'd make a separate branch (and PR) for any improvements unrelated to this one.

**Comment by doskabouter** [comment]
Ok, will do

**Pull request merged**